### DISCUSSION TITLE
DMMF and computed fields

### DISCUSSION THREAD

**Discussion by ch-arslan-kip-tm** [prisma/prisma#24362] [Q&A] [answered]
### Question

Using the [computed fields](https://www.prisma.io/docs/orm/prisma-client/queries/computed-fields) feature we can extend Prisma models with custom fields. However, those extensions are not reflected in the generated `Prisma.dmmf` object. Libraries like [PrismaSelect](https://paljs.com/plugins/select/) do make use of this object under the hood to filter out the incoming graphql fields that are not part of the model (in order to avoid errors for cases like dynamically resolved fields in graphql). However, because the computed fields are not included in the dmmf object, that library erroneously filters our valid computed fields.

To workaround the above-mentioned problem I have to monkey-patch the models fields in dmmf object. I was wondering if this is the only solution to this problem for now? Also, is it possible for 3rd party libraries to automatically get the metadata for computed fields, or is some work needed on the prisma-client side to enable such support?

**Comment by prkagrawal** [comment]
The issue you are facing with Prisma and the dmmf object not reflecting computed fields is a known limitation. Currently, Prisma's `dmmf` does not include custom fields added through extensions or computed fields. Consequently, tools like `PrismaSelect`, which rely on the `dmmf` for introspection, will not be aware of these fields and may incorrectly filter them out.

Potential Solutions and Workarounds :
- Monkey-Patching the `dmmf` Object: As you mentioned, one workaround is to monkey-patch the `dmmf` object to manually add the computed fields. This is a viable solution but has some drawbacks, such as maintenance overhead and the risk of breaking changes with future updates to Prisma.

- Custom Middleware: Implementing custom middleware in your GraphQL server to handle computed fields separately from PrismaSelect could be another approach. This middleware can manually include computed fields in the selection set before PrismaSelect processes the fields.

- Contributing to Third-Party Libraries: You could contribute to the libraries like `PrismaSelect` to include support for custom metadata. This would involve changes in how these libraries introspect the Prisma models and handle additional fields.

- Awaiting Prisma Enhancements

**Reply by ch-arslan-kip-tm** [reply]
Looks like an LLM response - lots of text, but no substantial information provided.

**Comment by ludralph** [comment, marked as answer]
Hi @ch-arslan-kip-tm 👋 

Thank you for raising this question.

The `Prisma.dmmf` object represents the Prisma Schema and is generated using `prisma generate` before any of your code is executed, which is why computed fields are not included. Computed fields are added by client extensions at runtime. Since `Prisma.dmmf` is a global variable and computed fields are specific to individual PrismaClient instances, it’s not feasible to reflect these dynamic changes in the dmmf. There could theoretically be multiple PrismaClient instances, each with different computed fields, complicating integration into a single dmmf. If you are comfortable using private APIs, the `_extensions` property on the PrismaClient instance might help.

**Reply by ch-arslan-kip-tm** [reply]
Thank you very much for the explanation, it does make sense. Your reference of the `_extensions` property is much appreciated, will definitely have a look at it.
